internal/querycoord/query_coord.go

```diff
@@ -425,5 +429,164 @@ func (qc *QueryCoord) watchHandoffSegmentLoop() {
+			nodeID2SegmentInfos := make(map[int64][]*querypb.SegmentInfo)
+			segmentOffset := make(map[int64]int)
+			onlineNodeIDs := make([]int64, 0)
+			for nodeID := range onlineNodes {
+				nodeInfo, err := qc.cluster.getNodeInfoByID(nodeID)
+				if err != nil {
+					log.Warn("loadBalanceSegmentLoop: get node info from query node failed", zap.Int64("nodeID", nodeID), zap.Error(err))
+					delete(onlineNodes, nodeID)
+					continue
+				}
+
+				updateSegmentInfoDone := true
+				leastSegmentInfos := make([]*querypb.SegmentInfo, 0)
+				segmentInfos := qc.meta.getSegmentInfosByNode(nodeID)
+				for _, segmentInfo := range segmentInfos {
+					leastInfo, err := qc.cluster.getSegmentInfoByID(ctx, segmentInfo.SegmentID)
+					if err != nil {
+						log.Warn("loadBalanceSegmentLoop: get segment info from query node failed", zap.Int64("nodeID", nodeID), zap.Error(err))
+						delete(onlineNodes, nodeID)
+						updateSegmentInfoDone = false
+						break
+					}
+					leastSegmentInfos = append(leastSegmentInfos, leastInfo)
+				}
+				if updateSegmentInfoDone {
+					nodeID2MemUsageRate[nodeID] = nodeInfo.(*queryNode).memUsageRate
+					nodeID2MemUsage[nodeID] = nodeInfo.(*queryNode).memUsage
+					nodeID2TotalMem[nodeID] = nodeInfo.(*queryNode).totalMem
+					onlineNodeIDs = append(onlineNodeIDs, nodeID)
+					nodeID2SegmentInfos[nodeID] = leastSegmentInfos
+					segmentOffset[nodeID] = 0
+				}
+			}
+			log.Debug("loadBalanceSegmentLoop: memory usage rage of all online query node", zap.Any("mem rate", nodeID2MemUsageRate))
+			if len(onlineNodeIDs) <= 1 {
+				log.Warn("loadBalanceSegmentLoop: there are too few online query nodes to balance", zap.Int64s("onlineNodeIDs", onlineNodeIDs))
+				continue
+			}
+
+			// check which nodes need balance and determine which segments on these nodes need to be migrated to other nodes
+			memoryInsufficient := false
+			loadBalanceTasks := make([]*loadBalanceTask, 0)
+			for {
+				sort.Slice(onlineNodeIDs, func(i, j int) bool {
+					return nodeID2MemUsageRate[onlineNodeIDs[i]] > nodeID2MemUsageRate[onlineNodeIDs[j]]
+				})
+
+				// the memoryUsageRate of the sourceNode is higher than other query node
+				sourceNodeID := onlineNodeIDs[0]
+				dstNodeID := onlineNodeIDs[len(onlineNodeIDs)-1]
+				memUsageRateDiff := nodeID2MemUsageRate[sourceNodeID] - nodeID2MemUsageRate[dstNodeID]
+				// if memoryUsageRate of source node is greater then 90%, and the max memUsageDiff is greater than 30%
+				// then migrate the segments on source node to other query nodes
+				if nodeID2MemUsageRate[sourceNodeID] > Params.OverloadedMemoryThresholdPercentage ||
+					memUsageRateDiff > Params.MemoryUsageMaxDifferencePercentage {
+					// select the segment that needs balance on the source node
+					offset := segmentOffset[sourceNodeID]
+					segmentInfo := nodeID2SegmentInfos[sourceNodeID][offset]
+					req, err := generateLoadBalanceRequests(dstNodeID, segmentInfo, nodeID2MemUsage, nodeID2TotalMem)
```

Comment:
the error thrown here does'nt mean enough memory, maybe just already balanced.
I would suggest to return nil req if cluster is already balanced rather than throw error out

Reply:
If it had been balance, the above judgment would have been true, and the function  generateLoadBalanceRequests 
 would not be executed